File: src/main/java/com/freya02/bot/commands/slash/DeleteButtonListener.java
package com.freya02.bot.commands.slash;

import com.freya02.botcommands.api.components.Components;
import com.freya02.botcommands.api.components.InteractionConstraints;
import com.freya02.botcommands.api.components.annotations.JDAButtonListener;
import com.freya02.botcommands.api.components.event.ButtonEvent;
import com.freya02.botcommands.api.utils.EmojiUtils;
import net.dv8tion.jda.api.Permission;
import net.dv8tion.jda.api.entities.Emoji;
import net.dv8tion.jda.api.entities.User;
import net.dv8tion.jda.api.interactions.components.buttons.Button;

public class DeleteButtonListener {
	private static final String DELETE_MESSAGE_BUTTON_LISTENER_NAME = "DeleteButtonListener: deleteMessage";
	private static final Emoji WASTEBASKET = EmojiUtils.resolveJDAEmoji("wastebasket");

	@JDAButtonListener(name = DELETE_MESSAGE_BUTTON_LISTENER_NAME)
	public void onDeleteMessageClick(ButtonEvent event) {
		event.deferEdit().queue();

		event.getMessage().delete().queue();
	}

	public static Button getDeleteButton(User allowedUser) {
		return Components.dangerButton(DELETE_MESSAGE_BUTTON_LISTENER_NAME)
				.setConstraints(InteractionConstraints.ofUsers(allowedUser).addPermissions(Permission.MESSAGE_MANAGE))
				.oneUse()
				.build(WASTEBASKET);
	}
}

File: src/main/kotlin/com/freya02/bot/commands/slash/DeleteButtonListener.kt
package com.freya02.bot.commands.slash

import com.freya02.botcommands.api.components.Components
import com.freya02.botcommands.api.components.annotations.JDAButtonListener
import com.freya02.botcommands.api.components.event.ButtonEvent
import com.freya02.botcommands.api.core.annotations.Handler
import com.freya02.botcommands.api.utils.EmojiUtils
import net.dv8tion.jda.api.Permission
import net.dv8tion.jda.api.entities.UserSnowflake
import net.dv8tion.jda.api.interactions.components.buttons.Button
import net.dv8tion.jda.api.interactions.components.buttons.ButtonStyle

@Handler
class DeleteButtonListener {
    @JDAButtonListener(name = DELETE_MESSAGE_BUTTON_LISTENER_NAME)
    suspend fun onDeleteMessageClick(event: ButtonEvent, components: Components) {
        event.deferEdit().queue()
        event.hook.deleteOriginal().queue()
        components.deleteComponentsById(event.message.components.flatMap { it.actionComponents }.mapNotNull { it.id })
    }

    companion object {
        private const val DELETE_MESSAGE_BUTTON_LISTENER_NAME = "DeleteButtonListener: deleteMessage"
        private val WASTEBASKET = EmojiUtils.resolveJDAEmoji("wastebasket")

        fun Components.messageDeleteButton(allowedUser: UserSnowflake): Button {
            return persistentButton(ButtonStyle.DANGER, emoji = WASTEBASKET) {
                oneUse = true
                bindTo(DELETE_MESSAGE_BUTTON_LISTENER_NAME)

                constraints += allowedUser
                constraints += Permission.MESSAGE_MANAGE
            }
        }
    }
}
